Implemented readonly-/ramdb preloading

DrizztDoUrden opened the pull request:
If convolution perf and find dbs have not been started loading yet, handle constructor starts a separate thread that does that. When a db is requested in some library call like find, instead of just creating a new instance it first checks if preload has started and waits for its result if needed.

Commit:
Implemented readonly-/ramdb reloading

Commit:
Merge branch 'develop' into ddu/ram-db-preload

Commit:
format

Commit:
Merge branch 'ddu/ram-db-preload' of github.com:ROCm/MIOpen into ddu/ram-db-preload

Commit:
Fixed incorrect handle initialization

Commit:
Fixed test crashes on GetPreloadedDb

Commit:
Fixed crash

Commit:
format

Commit:
changed std::async capture to parameters

Commit:
changed task back to capture

Commit:
Merge remote-tracking branch 'origin/develop' into ddu/ram-db-preload

Commit:
fixed unused warning

Commit:
Merge remote-tracking branch 'origin/develop' into ddu/ram-db-preload

Commit:
format

Commit:
Merge branch 'develop' into ddu/ram-db-preload

Commit:
Added fusion dbs to the preload system

Commit:
Merge remote-tracking branch 'origin/develop' into ddu/ram-db-preload

CAHEK7 commented on src/include/miopen/db_preload.hpp:
```diff
@@ -0,0 +1,85 @@
+
+auto GetDbPreloadStates() -> DbPreloadStates&;
+
+template <class Db>
+inline auto GetPreloadedDb(const fs::path& path) -> std::unique_ptr<Db>
```
Since `Db` can only be `RamDB` and `ReadOnlyRamDb`, I think it makes sense to move the function body into `cpp` file and explicitly instantiate this function for those two types.
It doesn't seem like a performance critical code or may get any benefits from inlining (except a bloating binary and longer compilation time).

CAHEK7 commented on src/handle_common.cpp:
```diff
@@ -0,0 +1,124 @@
+#include <array>
+#include <mutex>
+
+namespace miopen {
+template <class Db>
+static void
+StartPreloadingDb(DbPreloadStates& states, DbKinds db_kind, const fs::path& path, bool is_system)
+{
```
Personally, I prefer unnamed namespaces over static. At least it creates explicit section of the file-local functions, classes, etc., and creates a uniform way to declare file-local types (which is not possible usnig static).
```suggestion
namespace {
template <class Db>
void
StartPreloadingDb(DbPreloadStates& states, DbKinds db_kind, const fs::path& path, bool is_system)
{
...
}
}

namespace miopen {
```

CAHEK7 commented on src/handle_common.cpp:
```diff
@@ -0,0 +1,124 @@
+
+    std::unique_lock<std::mutex> lock(states.mutex, std::defer_lock);
+
+    if(!lock.try_lock() || states.started_loading)
```
Can we make `started_loading` atomic and swap the condition?

CAHEK7 commented on src/include/miopen/db_preload.hpp:
```diff
@@ -0,0 +1,85 @@
+
+    std::unique_lock<std::mutex> lock{states.mutex};
+
+    if(!states.started_loading)
+        return nullptr;
```
See a comment about atomic `started_loading`.
We can return earlier before even try to lock the mutex. But I didn't check is it worth or not.

Commit:
Resolved comments

Commit:
format

Commit:
Merge branch 'develop' into ddu/ram-db-preload

Commit:
Rearranged code a little

Commit:
Fixed setting db preload started flag at inapropriate time and reverted using it as a shortcut

DrizztDoUrden commented:
@CAHEK7 I have reverted making the flag atomic and placing it before the mutex because it looses its original purpose xD

Commit:
Merge branch 'develop' into ddu/ram-db-preload

Commit:
More complicated atomic street magic

Commit:
Fixed warning

Commit:
More fixes to the shortcutting logic

Commit:
updated which bases are actually preloaded

Commit:
missing #endif

Commit:
fixed lock

Commit:
Merge branch 'develop' into ddu/ram-db-preload

Commit:
got rid of lambda capture just to be safe

Commit:
Moved static back inside of a function

Commit:
changed atomics to be relaxed

Commit:
merged remote

Commit:
Added a test

Commit:
changed set_value to set_value_at_thread_exit in the test

Commit:
some consts

Commit:
Added a test case with very basic usage and no synchronisation

Commit:
fixed missing dllexport warnings

Commit:
Added waiting for the last db preload thread on destroying last handle

Commit:
format

Commit:
Some cleanup

Commit:
Moved nolint to where it was before

Commit:
Changed waiting for remaining threads to be more explicit

Commit:
Don't need lock to start the thread

Commit:
moved error checking around a little

Commit:
format

Commit:
tidy fix

Commit:
Merge remote-tracking branch 'origin/develop' into ddu/ram-db-preload

Commit:
Updated db_preload to stop thread on static variable destructor and to stop ongoing prefetches rather than waiting for them

Commit:
Merge remote-tracking branch 'origin/develop' into ddu/ram-db-preload

Commit:
Merge branch 'develop' into ddu/ram-db-preload

Commit:
fixed redundant copy

Commit:
Merge remote-tracking branch 'origin/develop' into ddu/ram-db-preload

Commit:
Merge remote-tracking branch 'origin/develop' into ddu/ram-db-preload

Commit:
Renamed tests in db_preload.cpp

junliume committed:
Merge branch 'develop' into ddu/ram-db-preload

Commit:
Merge remote-tracking branch 'origin/develop' into ddu/ram-db-preload

Commit:
Fixed "leak" like in develop

Commit:
Some cleanup:
Changed DbPreloadStates to be stored as shared_ptr to prevent them being desctructed while still in use
Changed related stop_source to be used correctly
Added a mutex to DbPreloadStates dtor to prevent it from executing while it is still being initialized, moved related thread construction under the mutex as well

Commit:
format

Commit:
fixed test build

Commit:
Merge remote-tracking branch 'origin/develop' into ddu/ram-db-preload

Commit:
added exception handling to db preloading

Commit:
fixed build

Commit:
format

Commit:
Merge remote-tracking branch 'origin/develop' into ddu/ram-db-preload

BradPepersAMD commented:
MIOpen is moving to the monorepo and old PR's need to be closed for this to happen.  If needed, this PR can be re-opened in the new repo.
